cmd/jaas/cmd/migratemodel.go

```diff
@@ -235,3 +269,108 @@ func (c *migrateModelCommand) getMigrationSpec(token string, modelUUID string) (
 		TargetUser: accountDetails.User,
 	}, nil
 }
+
+// validateUserMapping checks that the user mapping contains all necessary users
+// that have access to the model and its offers.
+func (c *migrateModelCommand) validateUserMapping(userMapping map[string]string, modelUUID, modelName string, jujuAPI MigrateAPI) error {
+	modelInfo, err := jujuAPI.ModelInfo([]names.ModelTag{names.NewModelTag(modelUUID)})
+	if err != nil {
+		return fmt.Errorf("could not get model info: %v", err)
+	}
+	modelUsers := modelInfo[0].Result.Users
+
+	var missingUserMessages []string
+
+	for _, user := range modelUsers {
+		// Skip checks for the "everyone" since the user
+		// shouldn't need to map this special case user.
+		if user.UserName == ofganames.EveryoneUser {
+			continue
+		}
+		if _, ok := userMapping[user.UserName]; !ok {
+			missingUserMessages = append(missingUserMessages, fmt.Sprintf("expected user %q who has %s access to the model", user.UserName, user.Access))
+		}
+	}
+
+	// To list the model offers, we need the model name and owner as unfortunately
+	// the model UUID is not sufficient to query the offers.
+	unqualifiedModelName, ownerTag, err := jujuclient.SplitModelName(modelName)
+	if err != nil {
+		return err
+	}
+	filter := crossmodel.ApplicationOfferFilter{
+		OwnerName: ownerTag.Id(),
+		ModelName: unqualifiedModelName,
+	}
+	offers, err := jujuAPI.ListOffers(filter)
+	if err != nil {
+		return fmt.Errorf("could not list application offers: %v", err)
+	}
+	for _, offer := range offers {
+		for _, user := range offer.Users {
+			if user.UserName == ofganames.EveryoneUser {
+				continue
+			}
+			if _, ok := userMapping[user.UserName]; !ok {
+				missingUserMessages = append(missingUserMessages, fmt.Sprintf("expected user %q who has %s access to offer %q", user.UserName, user.Access, offer.OfferName))
+			}
+		}
+	}
+	if len(missingUserMessages) > 0 {
+		return fmt.Errorf("user mapping is missing the following users:\n%s\n", strings.Join(missingUserMessages, "\n"))
+	}
+
+	return nil
+}
+
+func (c *migrateModelCommand) newJujuClient() (MigrateAPI, error) {
```

Comment:
This feels like it should be inherited rather than on the command, perhaps we wrap modelcmd.ControllerCommandBase.